### ISSUE TITLE
WebSocket connection error with Apple Safari 

### ISSUE THREAD

**Issue by pontscho** [lpereira/lwan#345]
Hi,

first of all thanks for clarifying my previous patch about WebSocket ping.

I found and another not-too-fun "issue" with Lwan's web socket implementation, which isn't a real bug in Lwan but in Apple Safari. For some unknown reasons it isn't send `Upgrade` header when it starts a new connection but if I comment out lines from `lwan-request.c:1344-1347`:
```
     const char *upgrade = lwan_request_get_header(request, "Upgrade");
     if (UNLIKELY(!upgrade || !streq(upgrade, "websocket")))
         return HTTP_BAD_REQUEST;
```
 helps establish web socket connection with Lwan without any other problem. 

Commenting out these lines is just a workaround, not a real solution but helps to interoperatibihity between lwan and Safari and strict check isn’t get injured because the next few lines are used to check the websocket connection. If the sec-websocket-key field is missing, it either means a total error on any ways.

What is your opinion about this case ?

pontscho

**Comment by lpereira** [comment]
This is pretty curious. Are you testing it directly or over some proxy? Have you looked at the raw headers before we parse them?

The Upgrade method is necessary for websockets according to the RFC, so I don't really know if we should remove the check for it. Have you tried other websockets implementations?

**Comment by lpereira** [comment]
Does it have a Sec-WebSocket-Version header? Is the value different than what other clients send?

Do other servers work with Safari even without this header?

**Comment by pontscho** [comment]
_Does it have a Sec-WebSocket-Version header? Is the value different than what other clients send?_

Yes, with a correct value. It had every necessary headers except the Upgrade.

_Do other servers work with Safari even without this header?_

I used nginx before for tls termination and it worked but I overwrote http headers with it. I will check it again.

**Comment by lpereira** [comment]
I just tried the websocket demo with Safari 15.4 here and ot seems to work! The Upgrade header is sent as expected. Maybe you're using an old version?

I checked other servers, like:

- gorilla/websocket
- libwebsockets
- tungstenite-rs

And all of these seem to require the presence of this header.

Now, if removing the check is needed for interoperability, then that's probably OK, although I'd prefer working around broken clients if at all possible.

**Comment by pontscho** [comment]
_I just tried the websocket demo with Safari 15.4 here and ot seems to work!_

I've tested it with Safari 13.1.4, it's quite old but significant amount of users use have this version. I've tested it again with other sites that uses websocket connection and that header is missing. Apple quality as well ... :)

_Now, if removing the check is needed for interoperability, then that's probably OK, although I'd prefer working around broken clients if at all possible._

Thanks.
